### ISSUE TITLE
Should lambdifying Max to numpy use np.amax or np.maximum?

### ISSUE THREAD

**Issue by vperic** [sympy/sympy#11027]
When lambdifying `Max` to the `numpy` format, `np.amax` is used. This function returns the maximal element of the passed array. While this is the same as the definition of sympy.Max ("Return, if possible, the maximum value of the list."), it doesn't strike me as very useful. This was added back in 2008 in commit `8f2653b0`. Numpy also has the function `maximum(a,b)`, which returns the element-wise maximum of two passed arrays. This seems like a more useful feature to have.

Should we lambdify to np.maximum instead of np.amax? (the same applies for Min, of course)

**Comment by asmeurer** [comment]
Note that Max takes multiple arguments whereas maximum only takes two, but max and min are associative so you can nest them.

**Comment by richardotis** [comment]
Reading the documentation for [`Max`](http://docs.sympy.org/0.7.1/modules/functions.html#sympy.functions.elementary.miscellaneous.Max), I'm not sure lambdifying to [`np.amax`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.amax.html) is even correct. It makes more sense (to me, anyway) to imagine users wanting to apply a "max" operation to some number of arguments and have it vectorized over the appropriate dimensions. That's the way the logical operators work in `NumPyPrinter`. We should probably always return [`np.maximum`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.maximum.html#numpy.maximum) and do nesting for more than two arguments, like @asmeurer says, something like

```python
def _print_Max(self, expr):
    "Max printer"
    if len(expr.args) == 1:
        return self._print(expr.args[0])
    result = 'maximum({0}, {1})'.format(self._print(expr.args[-2]),
                                        self._print(expr.args[-1]))
    for arg in reversed(expr.args[:-2]):
        result = 'maximum({0}, {1})'.format(self._print(arg), result)
    return  result
```

**Comment by bjodah** [comment]
@vperic to the best of my knowledge it didn't work until I wrote gh-10442 this year (motivated by a real-world need so I would argue it is very useful).

@vperic and @richardotis, I fail to see why `sympy.Max` would map to `numpy.maximum`:

```
>>> import numpy as np
>>> np.maximum([np.pi, np.exp(1)], [3, 3])
array([ 3.14159265,  3.        ])
>>> from sympy import Max, pi, exp
>>> Max([pi, exp(1)], [3, 3])
Traceback (most recent call last)
 ...
ValueError: The argument '[pi, E]' is not comparable.
```

whereas the amax case goes like:
```
>>> import numpy as np
>>> np.amax([np.pi, np.exp(1), 1])
3.1415926535897931
>>> from sympy import Max, pi, exp
>>> Max([pi, exp(1), 1])
π
```

Why not introduce `Maximum` in sympy, make it play nice with [NDimArray](https://github.com/sympy/sympy/blob/master/sympy/tensor/array/ndim_array.py#L6), and make it lambdify to `numpy.maximum`?

**Comment by asmeurer** [comment]
@bjodah that's the case with every lambdified function. `sympy.cos([1, 2])` doesn't work but `lambdify(x, cos(x), 'numpy')([1, 2])` does work.

I'd be curious to hear more about your use-case.

**Comment by bjodah** [comment]
@asmeurer, you're right my comparison was not fair. Perhaps @richardotis solution is the way to go given it does reproduce the behaviour of `lambdify` when `module='sympy'` for scalar input (which it looks as if it does), but our tests would guard against that anyways.

The usecasse was for obtaining bounds for the solution vector of a system of non-linear equations resulting from chemical equilibria. I use SymPy to generate the system of equations, see e.g. [this](http://hera.physchem.kth.se/~chempy/branches/master/examples/ammonical_cupric_solution.html) for a similar example.

**Comment by richardotis** [comment]
This is a bit off-topic but since @bjodah is developing chempy and is from KTH (birthplace of computational thermo), he may be very interested to check out my related project, [pycalphad](https://github.com/pycalphad/pycalphad), which also uses sympy under the hood to compute equilibrium, but for (primarily) inorganic materials.

**Comment by bjodah** [comment]
@richardotis thank you for reaching out, I'll definitely look into your project. Some of the things we are donig are very similar, feel free to contact me on my email too.

**Comment by rlloretb** [comment]
Any news on this issue?

**Comment by bjodah** [comment]
@rlloretb I think it's "only" a matter of someone writing a pull request together with a bunch of tests. Especially tests that highlights any differences in behavior between `amax` which is currently generated and `maximum`.

There have been some pull requests (see e.g. https://github.com/sympy/sympy/pull/25211) that simply change the code generated but without any written analysis of potential backwards compatiblity (nor any added tests).

Depending on how big the backcomp issue is, we can choose to either:
1. change the emitted code in the next release (if it's very unlikely to break any user's code)
2. have a deprecation cycle, perhaps using some option (e.g. "numpy_min_max=Auto|amin_amax|minimum_maximum"), and have the default be Auto and document in the release notes that the meaning of Auto will change from amin_amax to minimum_maximum in release sympy-1.x.

The exact implementation details of the deprecation cycle (and naming) is just an example, and perhaps its not even needed.

**Comment by bjodah** [comment]
fixed in #27483
